### PR TITLE
Refactor particle systems

### PR DESCRIPTION
**Pull request by cremebrule**

### CODE REVIEW

File: omnigibson/systems/micro_particle_system.py [hunk 1]
```diff
@@ -1042,21 +1131,35 @@ def generate_particle_instancer_on_object(
...
+        return inst
 
     @classmethod
     def remove_particle_instancer(cls, name):
         """
         Removes particle instancer with name @name from this system.
 
+        Note that if this is the default particle instancer (instancer 0), it will NOT delete this instancer,
+        but remove all of its owned particles
+
```
Comment (ChengshuLi) [inline omnigibson/systems/micro_particle_system.py:1162]: Is your code actually doing that? Maybe it should raise an error if someone is trying to remove the default particle instancer? Also, what does `instancer.remove` do?

File: omnigibson/systems/micro_particle_system.py [hunk 2]
```diff
@@ -1116,11 +1222,15 @@ def _sync_particle_instancers(cls, idns, particle_groups, particle_counts):
...
-                logging.debug(f"Got mismatch in particle instancer {name} when syncing, deleting and recreating instancer now.")
-                # Add this instancer to both the delete and creation pile
-                instancers_to_delete.add(name)
-                instancers_to_create.add(name)
+            if instancer.particle_group != info["group"]:
+                instancer.particle_group = info["group"]
+            count_diff = info["count"] - instancer.n_particles
+            if count_diff > 0:
+                # We need to add more particles to this group
+                instancer.add_particles(positions=np.zeros((count_diff, 3)))
```
Comment (ChengshuLi) [inline omnigibson/systems/micro_particle_system.py:1230]: Who is responsible for setting these particles to the correct position? load_state?

File: omnigibson/systems/micro_particle_system.py [hunk 3]
```diff
@@ -1132,6 +1242,9 @@ def _sync_particle_instancers(cls, idns, particle_groups, particle_counts):
             info = idn_to_info_mapping[idn]
             cls.generate_particle_instancer(idn=idn, particle_group=info["group"], n_particles=info["count"])
 
+        # Update the max particle instancer ID
+        cls.update_max_instancer_idn()
```
Comment (ChengshuLi) [inline omnigibson/systems/micro_particle_system.py:1246]: This line should be put in `generate_particle_instancer`.

File: omnigibson/object_states/contact_fluids.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            # If no links are specified, then we assume checking contact with any link owned by this object
+            # Otherwise, we check for exact match of link name
+            if (link is None and base == self.obj.prim_path) or (link is not None and link_name == body):
+                # Add to contacts and terminate early
+                contacts[inst].add(idx)
+                continue_traversal = False
+            return continue_traversal
+
+        # Grab the relaxed AABB of this object or its link for coarse filtering of particles to ignore checking
+        lower, upper = self.obj.states[AABB].get_value() if link is None else self.obj.links[link].aabb
```
Comment (ChengshuLi) [inline omnigibson/object_states/contact_fluids.py:43]:
```
link.prim_path.split("/")[-1]
self.obj.links[link_name]
```
should be like this?

File: omnigibson/object_states/covered.py [hunk 5]
```diff
@@ -33,7 +34,31 @@ def __init__(self, obj):
     @staticmethod
     def get_dependencies():
         # AABB needed for sampling visual particles on an object
-        return RelativeObjectState.get_dependencies() + [AABB]
+        return RelativeObjectState.get_dependencies() + [AABB, ContactFluids]
+
+    def reset(self):
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/object_states/covered.py:39]: suggestions:
1. only controllable objects (e.g. robots) have a reset function
2. StatefulObject should rename `reset` to `clear_cache` and call it in `load_state`. 
3. remove object state reset functionalities. Only `clear_cache` is allowed.

File: omnigibson/object_states/fluid_sink.py [hunk 6]
```diff
@@ -45,20 +45,18 @@ def _update(self):
...
-        for name in names_to_remove:
-            self.fluid_system.remove_particle_instancer(name)
+            # Get distances
+            idxs_to_remove = np.where(np.linalg.norm(particle_pos - fluid_sink_position.reshape(1, 3),
+                                           axis=-1) < self.max_sink_distance)[0]
+            inst.remove_particles(idxs=idxs_to_remove)
+
+            # Possibly remove this instancer if there are no particles left and this is not the default instancer
+            if inst.n_particles == 0 and inst != self.fluid_system.default_particle_instancer:
+                self.fluid_system.remove_particle_instancer(name)
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/object_states/fluid_sink.py:59]: suggestion: remove this

File: omnigibson/systems/micro_particle_system.py [hunk 7]
```diff
@@ -550,16 +601,32 @@ def initialize(cls, simulator):
...
         return 1 + 3 * len(cls.particle_instancers) + sum(inst.state_size for inst in cls.particle_instancers.values())
 
+    @classproperty
+    def default_particle_instancer(cls):
+        """
+        Returns:
+            PhysxParticleInstancer: Default particle instancer for this particle system
+        """
+        # TODO: Clean up this logic, ideally FluidSystem stuff should exist here, but it breaks because ClothSystem doesn't use any of this logic!
+        raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/systems/micro_particle_system.py:628]: suggestion: should be a function of `PhysicalParticleSystem`.

File: omnigibson/systems/micro_particle_system.py (hunk 1 above)
Comment (ChengshuLi) [reply omnigibson/systems/micro_particle_system.py:1162]: conclusion: comment outdated

File: omnigibson/object_states/contact_fluids.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+import omnigibson as og
+from omnigibson.object_states.object_state_base import RelativeObjectState
+from omnigibson.object_states.aabb import AABB
+from omnigibson.object_states.kinematics import KinematicsMixin
+from omnigibson.systems import FluidSystem
+from omni.physx import get_physx_scene_query_interface
+
+
+class ContactFluids(RelativeObjectState, KinematicsMixin):
+
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/contact_fluids.py:16]: done!

File: omnigibson/object_states/fluid_sink.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/fluid_sink.py]: done!

File: omnigibson/systems/micro_particle_system.py [hunk 9]
```diff
@@ -146,23 +212,13 @@ def particle_group(self):
...
+    def particle_group(self, group):
         """
-        Sets this point instancer's (x,y,z) cartesian translation relative to its parent prim
-
         Args:
-            pos (3-array): (x,y,z) relative position to set this prim relative to its parent prim
+            group (int): Particle group this instancer belongs to
         """
-        self.set_attribute(attr="xformOp:translate", val=Gf.Vec3f(*(pos.astype(float))))
+        return self.set_attribute(attr="physxParticle:particleGroup", val=group)
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/micro_particle_system.py]: done!

File: omnigibson/systems/micro_particle_system.py [hunk 10]
```diff
@@ -171,7 +227,7 @@ def particle_positions(self):
             np.array: (N, 3) numpy array, where each of the N particles' positions are expressed in (x,y,z)
                 cartesian coordinates relative to this instancer's parent prim
         """
-        return np.array(self.get_attribute(attr="positions")) + self.position
+        return np.array(self.get_attribute(attr="positions"))# + self.position
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/micro_particle_system.py]: done!

File: omnigibson/systems/micro_particle_system.py [hunk 11]
```diff
@@ -184,8 +240,8 @@ def particle_positions(self, pos):
         """
         assert pos.shape[0] == self._n_particles, \
             f"Got mismatch in particle setting size: {pos.shape[0]}, vs. number of particles {self._n_particles}!"
-        pos = (pos - self.position).astype(float)
-        self.set_attribute(attr="positions", val=array_to_vtarray(arr=pos, element_type=Gf.Vec3f))
+        # pos = (pos - self.position).astype(float)
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/micro_particle_system.py]: done!

File: omnigibson/systems/micro_particle_system.py (hunk 1 above)
Comment (cremebrule) [reply omnigibson/systems/micro_particle_system.py]: removed!

File: omnigibson/systems/micro_particle_system.py (hunk 3 above)
Comment (cremebrule) [reply omnigibson/systems/micro_particle_system.py]: done!

File: omnigibson/object_states/contact_fluids.py (hunk 4 above)
Comment (cremebrule) [reply omnigibson/object_states/contact_fluids.py]: yes, done!

File: omnigibson/object_states/covered.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/covered.py:39]: done!

File: omnigibson/systems/micro_particle_system.py [hunk 12]
```diff
@@ -1042,21 +1131,35 @@ def generate_particle_instancer_on_object(
...
+                n_particles=n_particles,
+                positions=particle_positions,
+                self_collision=self_collision,
+                prototype_indices=prototype_indices,
+            )
+        else:
+            inst.add_particles(
+                positions=particle_positions,
+                prototype_indices=prototype_indices,
+            )
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/micro_particle_system.py:1153]: done!

File: omnigibson/systems/micro_particle_system.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/micro_particle_system.py:627]: done!

**Pull request merged**